overlord/ifacestate/handlers.go

```diff
@@ -1039,7 +1044,32 @@ func batchConnectTasks(st *state.State, snapsup *snapstate.SnapSetup, conns map[
 	if len(ts.Tasks()) > 0 {
 		ts.AddTask(setupProfiles)
 	}
-	return ts, nil
+	return ts, hasHooks, nil
+}
+
+// firstTaskAfterBootWhenPreseeding finds the first task to be run for thisSnap
+// on first boot after mark-preseeded task, this is normally install hook.
+func firstTaskAfterBootWhenPreseeding(thisSnap string, markPreseeded *state.Task) (*state.Task, error) {
+	var firstTaskAfterBoot *state.Task
+	if markPreseeded.Change() == nil {
+		return nil, fmt.Errorf("internal error: %s task not in change", markPreseeded.Kind())
+	}
+	for _, ht := range markPreseeded.HaltTasks() {
+		if ht.Kind() == "run-hook" {
+			var hs hookstate.HookSetup
+			if err := ht.Get("hook-setup", &hs); err != nil {
+				return nil, fmt.Errorf("internal error: cannot get hook setup: %v", err)
+			}
+			if hs.Hook == "install" && hs.Snap == thisSnap {
+				firstTaskAfterBoot = ht
+				break
```

Comment:
why not just `return ht, nil` here ? then you can drop the `if firstTaskAfterBoot == nil` below and always return `fmt.Errorf(...)` since this is the only way to exit this loop with non-nil error.

Reply:
Yep, simplified, thanks.